app/models/piece.rb

```diff
@@ -153,7 +153,24 @@ def is_obstructed?(x1, y1, x2, y2)
         return true
       end
     end
-  end 
+  end
+
+  def valid_move?(new_x, new_y)
+    if square_occupied?(new_x, new_y) && self.color === game.get_piece_at_coor(new_x, new_y).color
+        return false
+    elsif
+      self.is_obstructed?(new_x, new_y)
```

Comment:
I would put the obstruction check into the corresponding models of the pieces since not all of them will be using the method.. for example, the knight does not care about obstructions and valid_move? would throw an error, not helping it..